deploy/ray-operator/controllers/raycluster_controller.go

```diff
@@ -0,0 +1,238 @@
+// +kubebuilder:rbac:groups=apps,resources=statefulsets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps,resources=statefulsets/status,verbs=get;update;patch
+func (r *RayClusterReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	_ = r.Log.WithValues("raycluster", request.NamespacedName)
+	log.Info("Reconciling RayCluster", "cluster name", request.Name)
+
+	// Fetch the RayCluster instance
+	instance := &rayiov1alpha1.RayCluster{}
+	err := r.Get(context.TODO(), request.NamespacedName, instance)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		log.Error(err, "Read request instance error!")
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, ignoreNotFound(err)
+	}
+
+	log.Info("Print instance - ", "Instance.ToString", instance)
+
+	// Build pods for instance
+	expectedPods := r.buildPods(instance)
+
+	expectedPodNameList := mapset.NewSet()
+	expectedPodMap := make(map[string]corev1.Pod)
+	needServicePodMap := mapset.NewSet()
+	for _, pod := range expectedPods {
+		expectedPodNameList.Add(pod.Name)
+		expectedPodMap[pod.Name] = pod
+		if strings.EqualFold(pod.Labels[common.ClusterPodType], common.Head) {
+			needServicePodMap.Add(pod.Name)
+		}
+	}
+
+	log.Info("Build pods according to the ray cluster instance", "size", len(expectedPods), "podNames", expectedPodNameList)
+
+	runtimePods := corev1.PodList{}
+	if err = r.List(context.TODO(), &runtimePods, client.InNamespace(instance.Namespace), client.MatchingLabels{common.RayClusterOwnerKey: request.Name}); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	runtimePodNameList := mapset.NewSet()
+	runtimePodMap := make(map[string]corev1.Pod)
+	for _, runtimePod := range runtimePods.Items {
+		runtimePodNameList.Add(runtimePod.Name)
+		runtimePodMap[runtimePod.Name] = runtimePod
+	}
+
+	log.Info("Runtime Pods", "size", len(runtimePods.Items), "runtime pods namelist", runtimePodNameList)
+
+	// record pod need to be deleted
+	difference := runtimePodNameList.Difference(expectedPodNameList)
+
+	// fill replicas with runtime if exists or expectedPod if not exists
+	var replicas []corev1.Pod
+	for _, pod := range expectedPods {
+		if runtimePodNameList.Contains(pod.Name) {
+			replicas = append(replicas, runtimePodMap[pod.Name])
+		} else {
+			replicas = append(replicas, pod)
+		}
+	}
+
+	// create service for head
+	if needServicePodMap.Cardinality() > 0 {
+		for elem := range needServicePodMap.Iterator().C {
+			podName := elem.(string)
+			svcConf := common.DefaultServiceConfig(*instance, podName)
+			rayPodSvc := common.ServiceForPod(svcConf)
+			if errSvc := r.Create(context.TODO(), rayPodSvc); errSvc != nil {
+				if errors.IsAlreadyExists(errSvc) {
+					log.Info("Pod service already exist,no need to create")
```

Comment:
Should this be an error? Under what conditions would this be called?

Reply:
For example, the node of head pod crash, the head pod will be evicted to another node, the service needn't to be created again.

Reply:
I will also add a warning event to the object